docs/source/specs/typespec/main.tsp

```diff
@@ -0,0 +1,251 @@
+ */
+model NoContentResponseResourceDeleted{
+    @statusCode
+    statusCode: 201,
+}
+
+namespace Errors {
+    @error
+    model ErrorMessage400 {
+        detail?: string = "JSON parse error - Expecting value: line 1 column 1 (char 0)",
+        status?: string = "404",
+        source?: string = "detail"
+    }
+
+    @error
+    model ErrorMessage403 {
+        detail?: string = "You do not have permission to perform this action.",
+        status?: string = "403",
+        source?: string = "detail"
+    }
+
+    @error
+    model ErrorMessage404 {
+        detail?: string = "Not found.",
+        status?: string = "404",
+    }
+
+    @error
+    model ErrorBasicResponse<STATUS>{
+        @statusCode 
+        statusCode: STATUS;
+    }
+
+    @error
+    model ErrorResponse400{
+        @statusCode
+        statusCode: 400,
+        errors: ErrorMessage400[]
+    }
+
+    @error
+    model ErrorResponse400AlreadyExist<T>{
+        @statusCode
+        statusCode: 400,
+        errors: T[]
+    }
+
+    @error
+    model ErrorResponse403{
+        @statusCode 
+        statusCode: 403,
+        errors: ErrorMessage403[]
+    }
+
+    @error
+    model ErrorResponse404{
+        @statusCode 
+        statusCode: 404,
+        errors: ErrorMessage404[]
+    }
+
+    @error
+    model ErrorResponse401{
+        @statusCode 
+        statusCode: 401,
+    }
+
+    scalar ErrorCode extends string;
+}
+
+interface CreateResource<BodyType> {
+    create<ModelType>(@body body: BodyType): CreatedResponseForModel<ModelType> | 
+    Errors.ErrorResponse401 |
+    Errors.ErrorResponse403 |
+    Errors.ErrorResponse400 |
+    Errors.ErrorBasicResponse<500>;
+}
+
+interface ListResource<T> {
+    list(...ListQueryParams): T[] |
+    Errors.ErrorResponse401 |
+    Errors.ErrorResponse403 |
+    Errors.ErrorBasicResponse<500>;
+}
+
+interface ReadResource<T> {
+    read(
+        @doc("Unique identification")
+        @path uuid: UUID;
+    ): T |
+    Errors.ErrorResponse401 |
+    Errors.ErrorResponse403 |
+    Errors.ErrorResponse404 |
+    Errors.ErrorBasicResponse<500>;
+}
+
+interface UpdateResource<B, R, E> {
+    update(
+        @doc("Unique identification")
+        @path uuid: UUID;
+        @body body: B
+    ): OkResponseForModel<R> |
+    Errors.ErrorResponse400AlreadyExist<E> |
+    Errors.ErrorResponse401 |
+    Errors.ErrorResponse403 |
+    Errors.ErrorBasicResponse<500>;
+}
+
+interface DeleteResource {
+    delete(
+        @doc("Unique identification")
+        @path uuid: UUID;
+    ): NoContentResponseResourceDeleted |
+    Errors.ErrorResponse401 |
+    Errors.ErrorResponse403 |
+    Errors.ErrorBasicResponse<500>;
+}
+
+@route("/workspaces")
+@tag("Workspaces")
+@doc("Operations about workspaces")
+namespace Workspaces {
+    model WorkspaceForRead {
+        @doc("Workspace A")
+        name: string = "Workspace A";
+        @doc("Description of Workspace A")
+        description?: string = "Description of Workspace A";
+    }
```

Comment:
Probably makes sense to include `parent?: UUID`.

We may want to change that to `parent` (not optional) later but will need to think through how tenants are registered in the API / backend.